include/sisl/fds/compact_bitset.hpp

```diff
@@ -0,0 +1,147 @@
+    static constexpr size_t word_size_bits() { return word_size_bytes() * 8; }
+    static constexpr uint64_t word_mask() { return bitword_type::bits() - 1; }
+
+public:
+    static constexpr bit_count_t inval_bit = std::numeric_limits< bit_count_t >::max();
+    static constexpr uint8_t size_multiples() { return word_size_bytes(); }
+
+    explicit CompactBitSet(bit_count_t nbits) {
+        DEBUG_ASSERT_GT(nbits, 0, "compact bitset should have nbits > 0");
+        nbits_ = s_cast< bit_count_t >(sisl::round_up(nbits, word_size_bits()));
+        size_t const buf_size = nbits_ / 8;
+
+        uint8_t* buf = new uint8_t[buf_size];
+        std::memset(buf, 0, buf_size);
+        s_ = r_cast< serialized* >(buf);
+        allocated_ = true;
+    }
+
+    CompactBitSet(sisl::blob const& buf, bool init_bits) : s_{r_cast< serialized* >(buf.bytes)} {
+        DEBUG_ASSERT_GT(buf.size, 0, "compact bitset initialized with empty buffer");
+        DEBUG_ASSERT_EQ(buf.size % word_size_bytes(), 0, "compact bitset buffer size must be multiple of word size");
+        nbits_ = buf.size * 8;
+        if (init_bits) { std::memset(buf.bytes, 0, buf.size); }
+    }
+
+    ~CompactBitSet() {
+        if (allocated_) { delete[] uintptr_cast(s_); }
+    }
+
+    bit_count_t size() const { return nbits_; }
+    void set_bit(bit_count_t start) { set_reset_bit(start, true); }
+    void reset_bit(bit_count_t start) { set_reset_bit(start, false); }
+
+    bool is_bit_set(bit_count_t bit) const {
+        bitword_type const* word_ptr = get_word_const(bit);
+        if (!word_ptr) { return false; }
+        uint8_t const offset = get_word_offset(bit);
+        return word_ptr->is_bit_set_reset(offset, true);
+    }
+
+    bit_count_t get_next_set_bit(bit_count_t start_bit) const { return get_next_set_or_reset_bit(start_bit, true); }
+    bit_count_t get_next_reset_bit(bit_count_t start_bit) const { return get_next_set_or_reset_bit(start_bit, false); }
+
+    void set_reset_bit(bit_count_t bit, bool value) {
+        bitword_type* word_ptr = get_word(bit);
+        if (!word_ptr) { return; }
+        uint8_t const offset = get_word_offset(bit);
+        word_ptr->set_reset_bits(offset, 1, value);
+    }
+
+    bit_count_t get_next_set_or_reset_bit(bit_count_t start_bit, bool search_for_set_bit) const {
+        bit_count_t ret{inval_bit};
+
+        // check first word which may be partial
+        uint8_t const offset = get_word_offset(start_bit);
+        bitword_type const* word_ptr = get_word_const(start_bit);
+        if (!word_ptr) { return ret; }
+
+        uint8_t nbit{0};
+        bool found = search_for_set_bit ? word_ptr->get_next_set_bit(offset, &nbit)
+                                        : word_ptr->get_next_reset_bit(offset, &nbit);
+        if (found) { ret = start_bit + nbit - offset; }
+
+        if (ret == inval_bit) {
+            // test rest of whole words
+            bit_count_t current_bit = start_bit + (bitword_type::bits() - offset);
```

Comment:
can we add a comment here by saying current_bit is aways the start bit of next word?
Maybe `start_bit_cur_word` would be more readable?

Reply:
I think the underlying reason is to just utilize what is there in bitset.hpp and didn't want to reimplement the entire logic again. If we perhaps reimplement them, we might do differently, but I prefer using battle tested piece and didn't want to spend time debugging and testing them.